Add an org-level CONTRIBUTING.md doc for Charmed-HPC (#19)
charmed-hpc/.github: CONTRIBUTING.md

maintainers quickly fix issues.

## Enhancement Proposals

The Charmed HPC core developers may already know what they want to add to Charmed HPC,
but they are always open to new ideas and potential improvements. GitHub Discussions is
a good place for discussing open-ended questions that pertain to the entire Charmed HPC project, but more focused enhancement proposal discussions can start within an issue tracker.

Please note that not all proposals may be incorporated into Charmed HPC. Also, please
know that spamming the maintainers to incorporate something you want will not improve the likelihood of being implemented; it may result in you receiving a temporary ban.

## Guidelines and Resources

The following guidelines should be followed if you are writing code to be merged into any of the respositories within Charmed HPC. Some repositories will have additional guidelines; please read the corresponding CONTRIBUTING.md within the repository of interest.

### Documentation

* Any code contribution should be accompanied by a correspsonding documentation contribution to [Charmed HPC's documentation](https://github.com/charmed-hpc/docs).

>   * If the code contribution does not require a documentation addition or change, please provide justification in the Pull Request.

Reviewer: [question]: Do you think it's worth mention diátaxis here? Or should the be scoped strictly to the `docs` repo's CONTRIBUTING.md?

### Conventional Commits 

* Follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)

### Juju and charmed operators

* Adhere to the operator development best practices outlined in the [operator development styleguide](https://juju.is/docs/sdk/styleguide).

### Python

> * Adhere to the Python code style guidelines outlined in [Python Enhancement Proposal 8](https://pep8.org/).
> 
> * Adhere to the Python docstring conventions outlined in
> [Python Enhancement Proposal 257](https://www.python.org/dev/peps/pep-0257/).
>   * *Docstrings must follow the
>   [Google docstring format](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings)*.

Reviewer: We could possibly expand this section to the following:

```suggestion
To have your Python code contributions accepted into Charmed HPC, you must:

- Follow the Python coding style guidelines outlined in [Python Enhancement Proposal 8](https://pep8.org/).
- Have all classes, functions, and methods include a docstring that documents their functionality. Docstrings must follow the [Google docstring format](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings).
- Use type annotations. Stub files can be used for providing type information for public modules.
- Pass all lint and static type checks.
- Pass all the repository's existing unit and integration tests.
- Add tests which prove the correctness of your changes.

If you are contributing to one of charmed operators used in Charmed HPC, you must:

- Follow the operator development best practices outlined in the [operator development styleguide](https://juju.is/docs/sdk/styleguide).
```

Goes more in-depth about our static typing practice, and the importance of passing all of the tests that we have defined. The "Add tests" bullet might need some wordsmithing however.

license

### Useful Resources 

The following list provides a variety of resources of revelance to Charmed HPC that may be worth skimming prior to making a contribution.

-  [Charmed HPC documentation](https://canonical-charmed-hpc.readthedocs-hosted.com/latest/)
-  [Juju SDK](https://juju.is/docs/sdk)
-  [Charmcraft documentation](https://canonical-charmcraft.readthedocs-hosted.com/stable/)

> -  [Ops library documentation](https://ops.readthedocs.io/en/latest/) for Juju

Author: Hmm ok, I phrased it that way since that's how that page does 'The `ops` library is a Python framework...'

* `cleanup` - Issues that pertain to improving the existing code base. Can also be used for pull requests.

* `feature` - Issues marked as an agreed upon feature to the code base. Can also be used for pull requests.

* `help wanted` - Issues where we need help from the greater open source community to solve.

* `good first issue` - Issues that may be a good fit for project newcomers looking to contribute.

For a complete look at a set of labels, see the
[slurm-charm labels page](https://github.com/charmed-hpc/slurm-charms/labels).

## Bug Reports

A bug is a *demonstrable problem* that is caused by errors in the corresponding repository's code.
Good bug reports make the code better, so
thank you for taking the time to report issues!

Guidelines for reporting bugs:

1. __Validate your issue__ &mdash; ensure that your issue is not being caused by either
a semantic or syntactic error in your environment.

2. __Use the GitHub issue search__ &mdash; check if the issue you are encountering has
already been reported by someone else.

3. __Check if the issue has already been fixed__ &mdash; try to reproduce your issue

> using the latest version of the Slurm charms.

Reviewer 2: ```suggestion
using the latest revision of the repository.
```

> 4. __Isolate the problem__ &mdash; the more pinpointed the issue with the Slurm charms
> is, the easier it is to fix.

Reviewer 2: ```suggestion
4. __Isolate the problem__ &mdash; the more pinpointed the issue, the easier it is to fix.
```
